src/V3Timing.cpp

```diff
@@ -0,0 +1,785 @@
+// - replace each delay with an await statement waiting on the global delay scheduler (with the
+// specified delay)
+// - for each event control:
+//     - if there's no corresponding trigger scheduler (see verilated_timing.h), create it
+//     - replace with an await statement waiting on the corresponding trigger scheduler
+// - for each fork, transform each statement to a separate suspendable function
+// - create additional necessary statements to inject into the global eval
+//
+//*************************************************************************
+
+#include "config_build.h"
+#include "verilatedos.h"
+
+#include "V3Ast.h"
+#include "V3EmitV.h"
+#include "V3Graph.h"
+#include "V3SenTree.h"
+#include "V3Timing.h"
+#include "V3UniqueNames.h"
+
+//######################################################################
+// Transform nodes affected by timing
+
+class TimingVisitor final : public VNVisitor {
+private:
+    // TYPES
+    using VarScopeSet = std::unordered_set<AstVarScope*>;
+    class DependencyVertex VL_NOT_FINAL : public V3GraphVertex {
+        AstNode* const m_nodep;  // AST node represented by this graph vertex
+    protected:
+        // ACCESSORS
+        virtual string name() const override {
+            return cvtToHex(nodep()) + ' ' + nodep()->prettyTypeName();
+        }
+        virtual FileLine* fileline() const override { return nodep()->fileline(); }
+        virtual string dotColor() const override { return nodep()->user2() ? "red" : "black"; }
+
+    public:
+        // CONSTRUCTORS
+        DependencyVertex(V3Graph* graphp, AstNode* nodep)
+            : V3GraphVertex{graphp}
+            , m_nodep{nodep} {}
+        virtual ~DependencyVertex() override = default;
+
+        // ACCESSORS
+        virtual AstNode* nodep() const { return m_nodep; }
+    };
+
+    // NODE STATE
+    //  AstNode::user1()                        -> bool.  Set true if node has been processed
+    //  Ast{NodeProcedure,CFunc,Fork}::user2()  -> bool.  Set true if process/task is suspendable
+    //  Ast{NodeProcedure,CFunc,Fork}::user3()  -> TimingDependencyVertex*.  Vertex in m_depGraph
+    VNUser1InUse m_inuser1;
+    VNUser2InUse m_inuser2;
+    VNUser3InUse m_inuser3;
+
+    // STATE
+    AstClass* m_classp = nullptr;  // Current class
+    AstScope* m_scopep = nullptr;  // Current scope
+    AstNode* m_procp = nullptr;  // NodeProcedure/CFunc/Fork we're under
+    bool m_underFork = false;  // Are we under a fork?
+    double m_timescaleFactor;  // Factor to scale delays by
+    const VTimescale m_timeprecision;  // Global time precision
+    V3UniqueNames m_procEnVarNames{"__VprocEn__"};  // Process enable var names
+    V3UniqueNames m_intraVarNames{"__Vintraval__"};  // Intra assign delay intermediate var names
+    V3UniqueNames m_dlyVarNames{"__Vdly__"};  // NBA intermediate var names
+    V3UniqueNames m_dlyChgVarNames{"__VdlyChg__"};  // NBA enable var names
```

Comment:
If it's an "Enable" then let's call it "En", not "Chg"

Reply:
No longer relevant